a quick fix for Issue:A indexing out of bounds issue shown in the global_4denvar regr…

TingLei-daprediction opened the pull request:
In response to Issue:" A indexing out of bounds issue shown in the global_4denvar regression test in setuprad.f90 #676" (https://github.com/NOAA-EMC/GSI/issues/676)  and following both online and off-line discussions, it is decided to retain the abi2km part in setuprad.f90 ( for potential future development/improvement), while a simple index-related checks are added to prevent the indexing error when GSI is built with debug mode. 
It is important to note that  this indexing error does not impact GSI results (when GSI is built with optimization options) since abi2km is currently hardwire consfigured to be not used. 

Fixes #676 

**DUE DATE for merger of this PR into `develop` is 2/22/2024 (six weeks after PR creation).**

TingLei-daprediction committed:
Issue:A indexing out of bounds issue shown in the global_4denvar regression test in setuprad.f90 #676

XuLu-NOAA reviewed (commented):
Hi, @TingLei-daprediction , is there a convention that the variables stay on the left side of the numbers in equation/condition? E.g. size(abi2km_bc) .ge. 2 instead of 2 .le. size(abi2km_bc).  Other than that I don't have questions about the fix itself.

TingLei-daprediction commented:
@XuLu-NOAA  when I have those lines,  kind of like I first have "if (i > size (abi2km_bc) ) then ",  where I replaced i with 2 , 3 and so on.   let me know if you prefer the form as "size(abi2km_bc) > 2" and so on.

RussTreadon-NOAA commented:
As @wx20jjung notes we need understand the reason for the `abi2km` code.   

Logical `abi2km` is local to `setuprad.f90` and is initialized to `.false.`
```
  abi        = obstype == 'abi'
  abi2km     = .false.

  ssmis=ssmis_las.or.ssmis_uas.or.ssmis_img.or.ssmis_env.or.ssmis
```
I do not see anywhere in `src/gsi` where `abi2km` is set to `.true.`.  I don't even see a mechanism by which `abi2km` can be set to `.true.` apart from changing the hardcoded value in `setuprad.f90`.  

Since `abi2km` is hardwired to `.false.`, the  code block in which `abi2km_bc` is used, namely,
```
           if (abi2km .and. regional) then                                                                                                                                  
              pred(:,i) = zero
              if (i>=2 .and. i<=4) then
                 if (tb_obs(i) > 190.0_r_kind .and. tb_obs(i) < 300.0_r_kind) then
                    pred(1,i)=1.0_r_kind
                    pred(2,i)=tb_obs(i)-abi2km_bc(i)
                    pred(3,i)=(tb_obs(i)-abi2km_bc(i))**2
                    pred(4,i)=(tb_obs(i)-abi2km_bc(i))**3
                 end if
              end if
           end if
```
will never be executed.   Hence Jim's question.  Is the `abi2km` code obsolete, orphaned, or in need of a rewrite?

If we are not ready to remove or rewrite code related to `abi2km` and `abi2km_bc`, we can restrict the initialization of `abi2km_bc` to when `abi2km` and `regional` are true.  That is, modify the initialization of `abi2km_bc` to
```
        if (abi2km .and. regional) then
           abi2km_bc = zero
           abi2km_bc(2) = 233.5_r_kind
           abi2km_bc(3) = 241.7_r_kind
           abi2km_bc(4) = 250.5_r_kind
        endif
```
Since `abi2km` is `.false.`, neither this block nor the block which uses `abi2km_bc` will be entered in regional or global gsi runs.

TingLei-daprediction commented:
@RussTreadon-NOAA . I can't answer the question if ab2km is obsolete or it will be updated/refreshed in the future.  That is why I retain them and give the current fix just to avoid indexing errors which would cause GSI abort  when GSI is built with debug mode. 
Second, if we use the conditions (abi2km.and.regional) , this would also prevent the abortion as reported for the global regression test as reported in the corresponding issue and leave further sorting out /treatment to the future if abi2km would be used in the future.   Do you prefer this fix?

JingCheng-NOAA commented:
This piece of code related to `abi2km` seems like someone's testing code to me. But in case it was designed for future use, I agree we can block this part for now. And I am more toward to @RussTreadon-NOAA 's solution to block this part at the initialization part. I have no more questions related to this fix.

JingCheng-NOAA commented:
> @RussTreadon-NOAA . I can't answer the question if ab2km is obsolete or it will be updated/refreshed in the future. That is why I retain them and give the current fix just to avoid indexing errors which would cause GSI abort when GSI is built with debug mode. Second, if we use the conditions (abi2km.and.regional) , this would also prevent the abortion as reported for the global regression test as reported in the corresponding issue and leave further sorting out /treatment to the future if abi2km would be used in the future. Do you prefer this fix?

@TingLei-daprediction then how about remove the `regional` in the condition, only use `abi2km`?

TingLei-daprediction commented:
@JingCheng-NOAA  I have no preference on this!

XuLu-NOAA commented:
Can we do both? The if condition will ensure the initialization is not used by default. But index check also resolves one potential issue if the abi2km is turned on in the future. Initialized or not, it is still a bug.

RussTreadon-NOAA commented:
My preference is to entirely remove the abi2km code because it won't be executed given how it is currently implemented in `setuprad.f90`.   

GSI source code is version controlled.  Removing abi2km code does not remove it from the repository.   A developer interested in the abi2km code can locate the PR at which it was removed via a search of github issues or PRs and, given this, checkout the GSI hash prior to the commit at which the code was removed.

TingLei-daprediction committed:
Revert "Issue:A indexing out of bounds issue shown in the global_4denvar regression test in setuprad.f90 #676"

This reverts commit 70ed687f72e85f74fcf3092b3e12b04e71ff0644.

TingLei-daprediction committed:
Issue:A indexing out of bounds issue shown in the global_4denvar regression test in setuprad.f90 #676, remove abc2km blocks

TingLei-daprediction commented:
@XuLu-NOAA  and @JingCheng-NOAA   So, since @RussTreadon-NOAA  has the same opinion as  @wx20jjung, one of the developers for this part, that this abc2km could be removed,  I have removed abc2km blocks and please review the current version.   Thanks.

ShunLiu-NOAA commented:
@JingCheng-NOAA  please review the new changes from Ting.

RussTreadon-NOAA reviewed (dismissed):
Thank you, @TingLei-daprediction , for removing unused code from `setuprad.f90`

Approve.

JingCheng-NOAA commented:
I have no more issues with the new changes.

On Fri, Jan 12, 2024 at 8:43 AM ShunLiu-NOAA ***@***.***>
wrote:

> @JingCheng-NOAA <https://github.com/JingCheng-NOAA> please review the new
> changes from Ting.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/GSI/pull/681#issuecomment-1889242757>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BAHEWIKCQL5DT6EZC3RFCOTYOE4XPAVCNFSM6AAAAABBWM2X4WVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOBZGI2DENZVG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ShunLiu-NOAA commented:
@TingLei-daprediction Could you please run regression test on HERA, WCOSS and ORION?

TingLei-NOAA commented:
@ShunLiu-NOAA  An update on the regression tests:  on hera, all passed, on orion, netcdf_fv3_regional failed for :
```
The memory for netcdf_fv3_regional_loproc_updat is 303768 KBs.  This has exceeded maximum allowable memory of 278040 KBs,
resulting in Failure memthresh of the regression test.
```
Considering this PR's property, it is safe to say this memory increase has nothing to with GSI itself and this regression test could be regarded as "pass" in my opinion.     The regression test results on wcoss2 are to be reported when it is available to me again.

TingLei-daprediction committed:
Issue:A indexing out of bounds issue shown in the global_4denvar regression test in setuprad.f90 #676, remove abc2km blocks
And Issue: AMSR2 Retrieval Predictor Caculation Went Wrong #678, added
test to abort with error message.

XuLu-NOAA reviewed (approved):
Thanks for adding these comments!

RussTreadon-NOAA commented on src/gsi/clw_mod.f90:
```diff
@@ -983,6 +983,12 @@ subroutine retrieval_amsr2(tb,nchanl,clw,kraintype,ierr)
+     .not.( (tb_use(11) -tb_use(12)) > 0_r_kind ) ) then
+     write(6,*)'The issue (GSI#678) needs further investigation and be resolved if this kind of obs is used' 
+     write(6,*)'Now, GSI would abort'
+     call stop2(72)
```
While I understand the rationale for this change (detect a protentional error), we do not want to merge this logic into `develop`.   

There are other places in `clw_mod` where a difference of two terms is used as the argument of `log`. Logic is in place to ensure the sign of the difference is greater than zero. 
 For example, lines 801 to 808 of `clw_mod.f90` in `develop` read
```
! ---------- Calculate predictors ---------------------------                                      

  if(tb_use(4) > tb_use(3)) then
     return
  endif

  pred_var_clw(1) = log(tb_use(3)-tb_use(4))  !(tb18v - tb18h)                                     
  pred_var_clw(2) = log(tb_use(6)-tb_use(7))   !(tb37v - tb37h)                                    
```
While this does not check for the possibility of a negative argument on the second `log`,  The first difference is checked.  This may be sufficient.  I am not familiar the characteristics of GMI channels.

Similar code could be placed before lines 993 and 994 and the above logic, write, and stop removed.

TingLei-daprediction replied:
@RussTreadon-NOAA  As some explanation at https://github.com/NOAA-EMC/GSI/issues/678, the problem is that we need more significant efforts to investigate the issue and the solution. For example, we don't know if 
```
if(tb_use(4) > tb_use(3)) then
     return
  endif
  ```  ( return to avoid the exception) is the one needed. It is not preferable because it won't give an warnming to the user/developer who might still think everything is going well.  That is why it is considered the current logics ( error message and stop) is needed before , in the future, when the data is to be considered and further work is applied on this issue.    That will be great if you share your opinions on that issue and more experts could discuss to decide a fix for being now.

RussTreadon-NOAA replied:
@TingLei-daprediction .  A similar argument could be made for other checks on the `log` argument found in `clw_mod.f90`.  These checks return execution or bypass `log` of an invalid value without writing to stdout and aborting `gsi.x` execution.  Yes, this means the invalid `log` argument goes undetected.   I assume the developers who wrote the code understood this and are OK with the given behavior.

The current write & abort change to `clw_mod.f90` in this PR is OK for testing but not for merger into `develop`.  If we want to add a check, I suggest mimicking the logic found in a similar section of `retrieval_gmi`.  

As @ADCollard explained we are not operationally processing AMSR2 data in the GSI.  @emilyhcliu and @xincjin-NOAA are looking into this.  I take this to mean that the GSI should not process AMSR2 data for the time being.

emilyhcliu replied:
@RussTreadon-NOAA You are correct.  We do not have the correct AMSR-2 BUFR from NESDIS yet.  In the meantime, we are using a few sample AMSR-2 BUFR data from GMAO for testing AMSR-2 in GSI.

TingLei-NOAA replied:
@RussTreadon-NOAA  "I assume the developers who wrote the code understood this and are OK with the given behavior." I am not sure if I understand what you meant.   The codes is always improving. It is obvious the "developer who wrote the code" would not be ok with the use of negative values in log function  in their  code(while it  exists to be deal with through this issue/PR) .

RussTreadon-NOAA replied:
@TingLei-daprediction , my comments are not meant to stop merger of this PR into GSI `develop`.  As stated above, I am not in favor of merging this type of print & abort into the authoritative GSI `develop`.  I am neither the GSI code manager nor the Handling Reviewer for this PR.  My comments do not preclude merger.

If the team decides to retain the proposed change to `clw_mod.f90` in this PR, please update the PR description to note that this PR resolves issue #678 in addition to resolving #676.

xincjin-NOAA replied:
I agree with @RussTreadon-NOAA's opinion.  At least GSI shouldn't be stopped in this situation because in the worst situation this type of satellite data, e.g., AMSR-2,  is not been used but there are many other Satellite data are still useful.  There is also another situation in which only a few points that are not valid  in the dataset. 

Print out message in this situation might also be not acceptable. Because this will print out tons of outputs saying one thing.

TingLei-daprediction committed:
revoke 'check and abort' change in clw_mod.f90

TingLei-daprediction commented:
@RussTreadon-NOAA  I had removed this change in question from this PR and will document our discussion in the issue 678 and let the decision be made/documented there.   Thanks.

RussTreadon-NOAA commented:
Thank you @TingLei-daprediction

TingLei-NOAA commented:
An update:  the regressions tests on WCOSS2 all passed!

RussTreadon-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/gsi_abi2km_index

RussTreadon-NOAA commented:
@ShunLiu-NOAA and @TingLei-daprediction 

With the merger of PR #670 into `develop` the changes to `setuprad.f90` in this PR conflict with the updated `develop`.  Resolution of the conflict is straight forward.   See Cactus `/lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr681` for a working copy of branch `feature/gsi_abi2km_index` with the conflict to `setuprad.f90` resolved.

@TingLei-NOAA , if you like I can attempt to push my updated working copy of `feature/gsi_abi2km_index` to your fork.  This may not work. I don't know.  I won't try unless you approve.

TingLei-NOAA commented:
@RussTreadon-NOAA  Thanks for taking care of this conflict.  I had sent the invitation to you as the collaborator on my fork and hence, after you accept it,  you could push your change to this PR.

RussTreadon-NOAA commented:
Conflict resolved @ [35b8aaf](https://github.com/NOAA-EMC/GSI/commit/35b8aaf15f371e7e204747bd4b54d5e98bf0aca2).

@TingLei-NOAA , peer reviewers need to re-approve this PR along with @ShunLiu-NOAA , the handling reviewer.

TingLei-NOAA commented:
@XuLu-NOAA @JingCheng-NOAA  your another round of reviewing for the updated PR will be appreciated.

ShunLiu-NOAA commented:
@TingLei-daprediction Could you please re-run regression test?

TingLei-NOAA commented:
@ShunLiu-NOAA  All tests passed the reproducibility  tests on hera, orion and wcoss2.
On each machine, a few tests failed for either exceeding time or memory , which are not considered issues for this PR. 
In conclusion,  this PR passes all regression tests on the above 3 machines.

ShunLiu-NOAA commented:
@TingLei-NOAA Thank you for completing regression test on WCOSS2, HERA and ORION.
